Source/Common/RenderShapeVisitor.cpp

```diff
@@ -9,55 +10,87 @@ using namespace Ashkal;
 using namespace std;
 
 namespace {
-  QPoint get_qt_coordinates(QWidget* widget, const Point& point) {
-    return QPoint(static_cast<int>(point.x),
-      widget->height() - static_cast<int>(point.y));
+  double mapTo(double value, double a1, double a2, double b1, double b2) {
+    return ((value - a1) * ((b2 - b1) / (a2 - a1))) + b1;
+  }
+
+  QPoint translateStageToView(const Point& point, const Camera& camera,
+      int width, int height) {
+    auto left = camera.get_pos().x - (camera.get_region().get_width() / 2);
+    auto right = camera.get_pos().x + (camera.get_region().get_width() / 2);
+    auto top = camera.get_pos().y - (camera.get_region().get_height() / 2);
+    auto bottom = camera.get_pos().y + (camera.get_region().get_height() / 2);
+    auto x = mapTo(point.x, left, right, 0, width - 1);
+    auto y = mapTo(height - point.y, top, bottom, 0, height - 1);
+    return QPoint(static_cast<int>(x), static_cast<int>(y));
   }
 }
 
-void Ashkal::render(const Shape& shape, const Point& point, QWidget* widget) {
+void Ashkal::renderShape(const Shape& shape, const Point& point,
+    const Camera& camera, QWidget* widget) {
   struct Visitor final : ShapeVisitor {
-    void operator ()(const Shape& shape, const Point& point, QWidget* widget) {
+    void operator ()(const Shape& shape, const Point& point,
+        const Camera& camera, QWidget* widget) {
       m_point = point;
       m_widget = widget;
+      m_camera = camera;
       shape.accept(*this);
     }
 
     void visit(const Ellipse& ellipse) override {
-      auto rendered_pos = Point{m_point.x + ellipse.get_major_radius(),
-        m_point.y + ellipse.get_minor_radius()};
-      auto pos = get_qt_coordinates(m_widget, m_point);
+      auto pos = translateStageToView(m_point, m_camera, m_widget->width(),
+        m_widget->height());
+      auto x_ratio = m_widget->width() / m_camera.get_region().get_width();
+      pos.setX(pos.x() - static_cast<int>(ellipse.get_x_radius() * x_ratio));
+      auto y_ratio = m_widget->height() / m_camera.get_region().get_height();
+      pos.setY(pos.y() - static_cast<int>(ellipse.get_y_radius() * y_ratio));
       auto painter = QPainter(m_widget);
-      painter.drawEllipse(pos, static_cast<int>(ellipse.get_major_radius()),
-        static_cast<int>(ellipse.get_minor_radius()));
-    }
+      painter.drawEllipse(pos.x(), pos.y(),
+        static_cast<int>(ellipse.get_x_radius() * x_ratio * 2),
```

Comment:
Constants always go in front.  You will always see ```2 * x``` as opposed to ```x * 2```.  ```2 * apple``` rather than ```apple * 2```. Generally left-to-right is more general to more specific, so in this case it should be ```2 * ratio * radius```.